### PR TITLE
Composite tiles

### PR DESCRIPTION
**Pull request by lilleyse**
For https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/23

Client-side implementation of composite tiles. We should wait on merging this until the server-side code is ready, as this introduces breaking changes to the tile formats.

* Added `byteLength` to all tile formats, added `batchSize` to `b3dm` tiles.
* Point clouds now define their bounding sphere from the tile's oriented bounding box.
* No longer using contentHeader.
* Moved a lot of the `ContentProvider` loading code from `request` to `init`. Now `request` only loads the url and passes the data along to `init`. Composite tiles will call `init` when creating inner tiles.

### CODE REVIEW

**Comment by pjcozzi** [comment]
This fails JSHint, https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/88820692

**Comment by pjcozzi** [comment]
> added batchSize to b3dm tiles.

Did we remove this from tiles.json?  Can we also remove `type` from tiles.json in this PR?

File: Apps/Sandcastle/gallery/Cities.html [hunk 1]
```diff
@@ -33,11 +33,12 @@
 var scene = viewer.scene;
 var city = scene.primitives.add(new Cesium.Cesium3DTileset({
     //url : 'http://localhost:8002/tilesets/Cambridge',
-    url : 'http://localhost:8002/tilesets/London_Canary_Wharf',
+    //url : 'http://localhost:8002/tilesets/London_Canary_Wharf',
```
Comment (pjcozzi) [inline Apps/Sandcastle/gallery/Cities.html]: Avoid submitting these temporary changes to Cities.html in the future.

File: Source/Core/OrientedBoundingBox.js [hunk 2]
```diff
@@ -606,25 +606,24 @@ define([
...
      *
-     * @param {OrientedBoundingBox} sphere The bounding box surrounding the occludee object.
-     * @param {Occluder} occluder The occluder.
-     * @returns {Boolean} <code>true</code> if the sphere is not visible; otherwise <code>false</code>.
+     * @param {OrientedBoundingBox} box The bounding box.
+     * @param {BoundingSphere} [result] The bounding sphere onto which to store the result.
+     * @returns {BoundingSphere} The computed bounding sphere.
      */
-    OrientedBoundingBox.isOccluded = function(box, occluder) {
+    OrientedBoundingBox.toBoundingSphere = function(box, result) {
```
Comment (pjcozzi) [inline Source/Core/OrientedBoundingBox.js]: Use `BoundingSphere.fromCornerPoints` or add a trivial `BoundingSphere.fromOrientedBoundingBox` (including tests, doc, and CHANGES.md), which could be a separate pull request into this branch and into master.

In general, we use `from` functions, not `to`.

File: Source/Scene/Points3DTileContentProvider.js [hunk 3]
```diff
@@ -64,66 +68,76 @@ define([
...
+        }).otherwise(function(error) {
             that.state = Cesium3DTileContentState.FAILED;
             that.readyPromise.reject(error);
+        });
+    };
+
+    Points3DTileContentProvider.prototype.init = function(arrayBuffer, byteOffset) {
+        byteOffset = defaultValue(byteOffset, 0);
+
+        var magic = getMagic(arrayBuffer, byteOffset);
```
Comment (pjcozzi) [inline Source/Scene/Points3DTileContentProvider.js]: Can you check if all the other tile formats, including the new composite, and verifying the magic number like this?

**Comment by pjcozzi** [comment]
> Moved a lot of the ContentProvider loading code from request to init. Now request only loads the url and passes the data along to init. Composite tiles will call init when creating inner tiles.

This is an OK approach.  If we stay with it, rename `init` to `initialize` to avoid the abbreviation.

However, can we do better?  Can we decouple the request and the promise management (which is the same-ish) from parsing the tile's arraybuffer (which is different per tile type).  For example, each tile's constructor would take an arraybuffer and offset, and would only be constructed once the request resolved.

File: Source/Scene/Composite3DTileContentProvider.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    "use strict";
+
+    /**
+     * DOC_TBA
+     */
+    var Composite3DTileContentProvider = function(tileset, tile, url, factory) {
+        this._url = url;
+        this._tileset = tileset;
+        this._tile = tile;
+        this._contentProviders = [];
```
Comment (pjcozzi) [inline Source/Scene/Composite3DTileContentProvider.js]: How do users of the Cesium API access this?  Can you update Cities.html to support mouse-over color change and so on (and also update it for i3dm, which I think will lead us to think more deeply about how to expose this in a uniform way).

**Comment by lilleyse** [comment]
> Did we remove this from tiles.json? Can we also remove type from tiles.json in this PR?

Both are removed from tiles.json.

> Can we decouple the request and the promise management (which is the same-ish) from parsing the tile's arraybuffer

The one obstacle I see for that is `ContentProvider` usually needs to load models, and can only send the ready promise when the model is loaded.

**Comment by pjcozzi** [comment]
>> Can we decouple the request and the promise management (which is the same-ish) from parsing the tile's arraybuffer

> The one obstacle I see for that is ContentProvider usually needs to load models, and can only send the ready promise when the model is loaded.

Ah, and the model resources, e.g., textures, may be separate files so the promise may resolve later.

I'm OK with sticking with your current approach.  As we add more tile formats, we'll see if it makes sense to change.

**Comment by pjcozzi** [comment]
Let me know when this is ready.

Do you plan to support a `tile.content.url` that points to another `tiles.json` file in this PR or a new one?

File: Source/Scene/Composite3DTileContentProvider.js (hunk 4 above)
Comment (lilleyse) [inline Source/Scene/Composite3DTileContentProvider.js]: I'm still not sure about this. I could just add a getter, but then it's tile.content.contents. Not that great. Maybe tile.content could sometime return an array? It's still odd.

The color changing works automatically though without any modifications.

Comment (pjcozzi) [inline Source/Scene/Composite3DTileContentProvider.js]: > The color changing works automatically though without any modifications.

Hmm.  Maybe we don't need to expose this for now since users are interacting with the inner tiles in the same way they interact with standalone tiles?  We could add a `TODO` and revisit later if needed.

Comment (pjcozzi) [inline Source/Scene/Composite3DTileContentProvider.js]: Just curious - did you test composites nested in composites?  That is test data we should have.

**Comment by lilleyse** [comment]
I'll do it in the same PR. I'll have that ready later.

The rest of the changes are ready for review. I also standardized the getMagic code and made some changes to `getStringFromTypedArray`.

File: Source/Scene/Batched3DModel3DTileContentProvider.js [hunk 5]
```diff
@@ -122,75 +118,95 @@ define([
...
-                byteOffset += batchTableLength;
-
-                // PERFORMANCE_IDEA: is it possible to allocate this on-demand?  Perhaps keep the
-                // arraybuffer/string compressed in memory and then decompress it when it is first accessed.
-                //
-                // We could also make another request for it, but that would make the property set/get
-                // API async, and would double the number of numbers in some cases.
-                batchTableResources.batchTable = JSON.parse(batchTableString);
-            }
+        // TODO : rename to batchLength?
```
Comment (pjcozzi) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: Let's rename to `batchTableLength`.

File: Source/Scene/Points3DTileContentProvider.js [hunk 6]
```diff
@@ -64,66 +68,77 @@ define([
...
+        if (version !== 1) {
+            throw new DeveloperError('Only Points tile version 1 is supported.  Version ' + version + ' is not.');
+        }
+        //>>includeEnd('debug');
+        byteOffset += sizeOfUint32;
+
+        // Skip byteLength
+        byteOffset += sizeOfUint32;
+
+        var numberOfPoints = view.getUint32(byteOffset, true);
```
Comment (pjcozzi) [inline Source/Scene/Points3DTileContentProvider.js]: Rename to `pointsLength`.

File: Source/Scene/Batched3DModel3DTileContentProvider.js (hunk 5 above)
Comment (lilleyse) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: What if the tile has no batch table? The terminology could be confusing.

Comment (pjcozzi) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: Then it would be zero.  This is consistent with 3D Tiles naming: https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/26

What do you suggest?

File: Source/Scene/Instanced3DModel3DTileContentProvider.js [hunk 7]
```diff
@@ -83,123 +81,136 @@ define([
...
+
+        //>>includeStart('debug', pragmas.debug);
+        if((gltfFormat !== 0) && (gltfFormat !== 1)) {
+            throw new DeveloperError('Only glTF format 0 (url) or 1 (embedded) are supported. Format ' + gltfFormat + ' is not');
+        }
+        //>>includeEnd('debug');
+
+        if (gltfFormat === 0) {
+            var gltfUrl = getStringFromTypedArray(gltfView);
+            var url = (new Uri(gltfUrl).isAbsolute()) ? gltfUrl : this._tileset.url + gltfUrl;
```
Comment (pjcozzi) [inline Source/Scene/Instanced3DModel3DTileContentProvider.js]: This should be `uri`, right?  Just like glTF uses `uri`, not `url`.

File: Source/Scene/Batched3DModel3DTileContentProvider.js (hunk 5 above)
Comment (lilleyse) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: But we still need the batch size even if it doesn't have a batch table. My comment was to rename `batchSize` to `batchLength`.

File: Source/Scene/Composite3DTileContentProvider.js (hunk 4 above)
Comment (lilleyse) [inline Source/Scene/Composite3DTileContentProvider.js]: I just tested this and it works.

File: Source/Scene/Instanced3DModel3DTileContentProvider.js (hunk 7 above)
Comment (lilleyse) [inline Source/Scene/Instanced3DModel3DTileContentProvider.js]: We talked about this earlier, but do we actually want to support data `uri`? That's sort of the point of embedding the glTF in the tile rather than referencing an external resource.

The same question applies to tiles.json which points to tiles by `url`. Should we support `uri` there too?

File: Source/Scene/Batched3DModel3DTileContentProvider.js (hunk 5 above)
Comment (pjcozzi) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: Yes, you are right.  Let's go with `batchLength` and `batchTableLength`.

File: Source/Scene/Instanced3DModel3DTileContentProvider.js (hunk 7 above)
Comment (pjcozzi) [inline Source/Scene/Instanced3DModel3DTileContentProvider.js]: I can't come up with a good reason to use a data uri instead of embedded binary glTF other than if a client only wanted to implement a subset of the i3dm spec (which we would not recommend).

The case for supporting uris in both places is consistency.  However, 3D Tiles only needs to be consistent with itself (not with glTF, which does uris) so I say we just stick with url.

**Comment by pjcozzi** [comment]
@lilleyse do you need to make any other updates to this PR before I do a final review?

**Comment by lilleyse** [comment]
I made the last couple of renaming changes. This is ready for another look.

File: Source/Scene/Cesium3DTileset.js [hunk 8]
```diff
@@ -165,22 +195,33 @@ define([
...
+                        var childTile = new Cesium3DTile(tileset, baseUrl, childHeader, t.cesium3DTile);
                         t.cesium3DTile.children.push(childTile);
 
                         stack.push({
                             header : childHeader,
                             cesium3DTile : childTile
                         });
+
+                        // Check if the tile's content url is a tiles.json file.
+                        if (defined(childHeader.content)) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Interesting approach.  I planned to do this by implementing a `3DTileContentProvider` with a `Cesium3DTileset` primitive, which would be a bit cleaner than this.  However, this approach will help with cache management, rendering budget, etc. since it isn't spread across multiple `Cesium3DTileset` primitives.  Let's see how it goes.

**Comment by pjcozzi** [comment]
Thanks again @lilleyse!
